### ISSUE TITLE
Signals are getting stale values inside async methods within a Suspense

### ISSUE THREAD

**Issue by LinqToException** [solidjs/solid#2180]
### Describe the bug

When running under `<Suspense>` in async functions, signal setters seem to revert/not keep the actual signal value. The example provided uses a `createResource` call which is setting a signal (which is only set in that function). Before the `await`, the signal works fine - after the `await` however, the signal seems to revert back to whatever state it had before. As pseudo-example code:

```ts
async () => {
  signal(); // Signal is 0
  setSignal(i => i + 1); // Signal is now 1
  await Promise.resolve(); // any async method (cb); in particular any function that does neither read nor set the signal
  signal(); // Signal is 0
  setSignal(i => i - 1); // Signal is now -1
};
```

I've tried to make a minimal example using the (limited) knowledge that I have. It's important to note that it seems to work fine when browsing directly to the affected page (i.e. the issue does not pop up and the setters behave as expected), but it happens whenever the component housing this logic is navigated to (either from another component, or from itself to another component an then back).

### Your Example Website or App

https://stackblitz.com/~/github.com/LinqToException/solidjs-signal-suspensethingy

### Steps to Reproduce the Bug or Issue

1. Using the sample provided, open the page on /bork. working() should report 0.
2. Go to Home (/)
3. Go back to /bork. Working should report -1.

The produced output should be similar to:

```
pre-inc: 0
inc working from 0 value 0
post-inc: 1
pre-cb: 1
post-cb: 0
pre-dec:  0
dec work from 0 value 0
post-dec -1
```

(note the change of the value pre/post cb).

### Expected behavior

I expect the signal to keep its value after `await`ing another method.

Expected output should be similar to:
```
pre-inc: 0
inc working from 0 value 0
post-inc: 1
pre-cb: 1
post-cb: 1
pre-dec:  1
dec work from 1 value 1
post-dec 0
```

### Platform

- OS: Windows
- Browser: Firefox
- Version: 126

### Additional context

The issue happens in pure client-side solid, i.e. there's no hydration or similar going on.

There are two workarounds that seem to "mitigate" this problem:

1.It does not occur if the component in question is not under a `<Suspense>`
2. It does not occur if the code querying or setting the signal post-`await` is executed delayed (`setTimeout(..., 0)`).

**Comment by ryansolid** [comment]
Thanks for the detailed report. I'm going to take a look. My guess is the router is involved and the router use Transitions on navigate which would keep the values in the past. But I will see if I can figure out what is going on.

**Comment by LinqToException** [comment]
Thank you for taking the time to analyze the issue.

As someone who's a more "casual" TS/JS developer, especially when it comes to reactivity, I find it a bit hard to follow some of those things you've mentioned however - as in, I suppose I can understand that this is an anti-pattern (as async and reactivity seem to be a bit awkward), but what would the "proper" pattern look like?

This is something that makes me feel a bit "left out in the rain" so to speak, especially when it comes to routing/resources. In this case, the basic idea for me was to have a `fetch`-operation that I could then query for the UI (i.e. `<input disabled={fetchIsRunning()} />`). The original code had a bool (with the same issues), but since I had multiple fetches that triggered the same lockout, I've moved it into a counter (as an attempt to circumvent this issue).

Suspense/Transitions itself do not necessarily work nicely for me in this case either, because they appear to be too wide-meshed (i.e. can't disable one processing form while offering the user to work on another one on the same page).
